delivery_carrier_preference/models/stock_move.py

```diff
@@ -41,6 +41,7 @@ def release_available_to_promise(self):
             existing_carriers[picking.id] = picking.carrier_id.id
 
             picking.add_preferred_carrier()
+            picking.group_id.carrier_id = picking.carrier_id
```

Comment:
I don't think that's correct, because if we grouped several SO in the same group, we'll update the method for all.
It seems we reach a shadow part of the ties between:

* stock_picking_group_by_partner_by_carrier
* stock_available_to_promise_release
* delivery_carrier_preference

I'm not sure delivery_carrier_preference can really be generic (also because of the weird rules it uses) maybe it should only act as a glue for our project.

Some thoughts, situation at this line:

* The grouping by partner and by carrier has been done already, they have the same procurement group
* Despite this, we are conditionally changing the group carrier, so it has an impact on the routing of the other pickings (but the other pickings are not updated with the same carrier, in any case we have an issue)
* The routing is applied the line after in `super().release_available_to_promise()`, the routing depends on the carrier (at least in our case), so we should have it on the group at this point
* Which means we should split the procurement group *here* if the carrier is different as the previous one
* We do not expect to have chained transfers before the release, so we should not have any other transfer to move in this group (?)
* In the current implementation, the procurement group is indeed split, but only after the routing is applied. I have the feeling that moving the split of the group before only could resolve the issue, but there might be a reason I miss having it after, maybe we need the release to be done for a reason?

Reply:
Rereading stock_picking_group_by_partner_by_carrier, I see I wasn't correct about what it does: it groups all the moves in the same picking, not different OUT pickings in the same group. We can have several sales order in the same group, but in the same OUT picking (+ backorders).

Reply:
So **if** we assume that we have only one OUT at a time for a procurement group, setting the carrier on the line may work.
However, this method is doing this:

1. Setting new carrier on group
2. apply release/routing
3. Loop on backorders and assign a new group to each if they had a different carrier

Which means we'll have one new group per backorder (even if we are not really supposed to have several at a time?)

It seems more simple to do this:

1. Create a new group for the current transfer if its carrier is different and the group has other transfers, set the carrier on the group
2. apply release/routing

Reply:
The problem is... before calling super, we don't know yet what will be released or not, and yet, we need to know and apply the carrier before releasing because that's what applies the routing.

Options: 
1. find a hook to place the carrier update
2. change the weight computation to take only in account the "available to promise" quantity